tuist/tuist
# Fix static products false positive lint warning

Resolves https://github.com/tuist/tuist/issues/973

### Short description 📝

- In the event a static product was linked more than once in a graph a warning was raised
- The `GraphLinter` had a step to check for this scenario during its graph traversal
- The previous method could accidentally flag false positives in the event multiple test targets depended on the same static product
- Additionally, depending on the order the entry nodes of a graph were evaluated there were scenarios which the check wouldn't catch real issues due to skipping evaluated nodes

### Solution 📦

- Add a dedicated `StaticProductsGraphLinter` component
- Evaluate each graph entry node separately to avoid flagging products linked multiple times when traversing from different entry nodes
- Cache the results for each node to flag issues regardless of the order nodes are visited for evaluation

### Implementation 👩‍💻👨‍💻

- [x] Create a dedicated linter `StaticProductsGraphLinter`
- [x] Update test to include more cases (multiple test targets, hosted test targets etc...)

### Test Plan 🛠

- run `tuist generate` within `fixtures/ios_app_with_static_frameworks`
- verify no warnings are displayed
- manually modify the fixture to introduce an error

e.g.

```swift
// Modules/D/Project.swift
import ProjectDescription

let project = Project(name: "D",
                      targets: [
                          Target(name: "D",
                                 platform: .iOS,
                                 product: .framework,
                                 bundleId: "io.tuist.D",
                                 infoPlist: "Info.plist",
                                 sources: "Sources/**",
                                 dependencies: [
                                       .project(target: "B", path: "../B") // <-- B is transitively linked by App
                          ])
])

```

- Run `tuist generate` and verify a warning is displayed:

```
Warning: - Target "B" has been linked against ["App", "D"], it is a static product so may introduce unwanted side effects.
```
